sensors/cmd/start.go

```diff
@@ -70,20 +45,194 @@ func Start() {
+
+	ctx := logging.WithLogger(signals.SetupSignalHandler(), logger)
+
+	manager := NewSensorContextManager(logger, kubeClient, dynamicClient, busConfig, ebSubject, hostname)
+	defer manager.Close()
 
-	logger = logger.With("sensorName", sensor.Name)
-	for name, value := range sensor.Spec.LoggingFields {
-		logger.With(name, value)
+	if err := manager.Start(ctx); err != nil {
+		logger.Fatalw("failed to start sensor context manager", zap.Error(err))
+	}
+
+	for {
+		select {
+		case f := <-manager.C:
+			if err := f(); err != nil {
+				logger.Fatalw("failed to listen to events", zap.Error(err))
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+type SensorContextManager struct {
+	C               chan func() error
+	logger          *zap.SugaredLogger
+	kubeClient      kubernetes.Interface
+	dynamicClient   dynamic.Interface
+	eventBusConfig  *eventbusv1alpha1.BusConfig
+	eventBusSubject string
+	hostname        string
+}
+
+func NewSensorContextManager(logger *zap.SugaredLogger, kubeClient kubernetes.Interface, dynamicClient dynamic.Interface, eventBusConfig *eventbusv1alpha1.BusConfig, eventBusSubject, hostname string) *SensorContextManager {
+	return &SensorContextManager{
+		C:               make(chan func() error),
+		logger:          logger,
+		kubeClient:      kubeClient,
+		dynamicClient:   dynamicClient,
+		eventBusConfig:  eventBusConfig,
+		eventBusSubject: eventBusSubject,
+		hostname:        hostname,
+	}
+}
+
+func (scm *SensorContextManager) Start(ctx context.Context) error {
+	// SENSOR_PATH env variable is used to indicate if this sensor definition
+	// has opted in to live reload
+	path, liveReload := os.LookupEnv("SENSOR_PATH")
+
+	switch liveReload {
+	case true:
+		return scm.startLiveReload(ctx, path)
+	case false:
+		return scm.startDefault(ctx)
+	}
+
+	return nil
+}
+
+func (scm *SensorContextManager) Close() {
+	close(scm.C)
+}
+
+func (scm *SensorContextManager) startDefault(ctx context.Context) error {
+	sensor, err := scm.readSensorFromEnv()
+	if err != nil {
+		scm.logger.Fatalw("failed to read sensor from env variable", zap.Error(err))
 	}
 
-	ctx := logging.WithLogger(signals.SetupSignalHandler(), logger)
 	m := metrics.NewMetrics(sensor.Namespace)
 	go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
 
-	logger.Infow("starting sensor server", "version", argoevents.GetVersion())
-	sensorExecutionCtx := sensors.NewSensorContext(kubeClient, dynamicClient, sensor, busConfig, ebSubject, hostname, m)
-	if err := sensorExecutionCtx.Start(ctx); err != nil {
-		logger.Fatalw("failed to listen to events", zap.Error(err))
+	sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+	scm.C <- func() error { return sensorCtx.Start(ctx) }
+
+	return nil
+}
+
+func (scm *SensorContextManager) startLiveReload(ctx context.Context, path string) error {
+	// watch for sensor file changes
+	watcher, err := fsnotify.NewWatcher()
```

Comment:
Is it possible that between the time the sensor is created and this watcher starts, someone can modify the sensor definition?

Reply:
I guess that's possible. Would that be an issue though? We would read the correct (most recent) information

Reply:
hmm should be fine them. watcher is created before we read from the file :+1: